main/src/ca/uwaterloo/flix/language/phase/Monomorph.scala

```diff
@@ -655,23 +658,50 @@ object Monomorph extends Phase[TypedAst.Root, TypedAst.Root] {
       // Lookup the definition and its declared type.
       val defn = root.defs(sym)
 
-      // Check if the function is non-polymorphic.
-      if (defn.tparams.isEmpty) {
-        return sym
-      }
+      specializeDef(defn, tpe)
+    }
+
+    /**
+      * Returns the def symbol corresponding to the specialized symbol `sym` w.r.t. to the type `tpe`.
+      */
+    def specializeSigSym(sym: Symbol.SigSym, tpe: Type): Symbol.DefnSym = {
+      val sig = root.sigs(sym)
+
+      // lookup the instance corresponding to this type
+      val instances = root.instances(sig.sym.clazz)
+
+      val defn = instances.flatMap {
+        inst => inst.defs.find {
+          defn => defn.sym.name == sig.sym.name && Unification.unifyTypes(defn.declaredScheme.base, tpe).isOk
+          // MATT probably need to do leq check instead
+        }
+      }.head // MATT do something monadic or throw internalcompilerexception if this is impossible
```

Comment:
I agree. At least at first sight, it seems that monomorphization can now fail, if constraints are un-satisfiable.

Reply:
After an update, I believe it should now be impossible for the number of matches to be different from one.